timesync/ticker.go

```diff
@@ -0,0 +1,180 @@
+	layerChannels   map[types.LayerID]chan struct{}
+	log             log.Log
+}
+
+func NewTicker(c Clock, lc LayerConverter) *Ticker {
+	return &Ticker{
+		subs:            newSubs(),
+		lastTickedLayer: lc.TimeToLayer(c.Now()),
+		clock:           c,
+		conv:            lc,
+		layerChannels:   make(map[types.LayerID]chan struct{}),
+		log:             log.NewDefault("ticker"),
+	}
+}
+
+var (
+	errNotStarted     = errors.New("ticker is not started")
+	errNotMonotonic   = errors.New("tried to tick a previously ticked layer")
+	errMissedTicks    = errors.New("missed ticks for one or more subscribers")
+	errMissedTickTime = errors.New("missed tick time by more than the allowed threshold")
+)
+
+const sendTickThreshold = 500 * time.Millisecond // allow up to sendTickThreshold
+
+// Notify notifies all the subscribers with the current layer
+// if the tick time has passed notify is skipped and errMissedTickTime is returned
+// notify may be skipped also for non-monotonic tick
+// if some of the subscribers where not listening, they are skipped. In that case, errMissedTicks is returned along the number of subscribers not listening
+func (t *Ticker) Notify() (int, error) {
+	if !t.started {
+		return 0, errNotStarted
+	}
+
+	t.m.Lock()
+
+	layer := t.conv.TimeToLayer(t.clock.Now())
+	// close prev layers
+	for l := t.lastTickedLayer + 1; l <= layer; l++ {
+		if layerChan, found := t.layerChannels[l]; found {
+			close(layerChan)
+			delete(t.layerChannels, l)
+		}
+	}
+
+	// the tick was delayed by more than the threshold
+	if t.timeSinceLastTick() > sendTickThreshold {
+		t.log.With().Warning("skipping tick since we missed the time of the tick by more than the allowed threshold", log.String("threshold", sendTickThreshold.String()))
+		t.m.Unlock()
+		return 0, errMissedTickTime
+	}
+
+	// already ticked
+	if layer <= t.lastTickedLayer {
+		t.log.With().Warning("skipping tick to avoid double ticking the same layer (time was not monotonic)",
+			log.Uint64("current", uint64(layer)), log.Uint64("last_ticked_layer", uint64(t.lastTickedLayer)))
+		t.m.Unlock()
+		return 0, errNotMonotonic
+	}
+	missedTicks := 0
+	t.log.Event().Info("release tick", log.LayerId(uint64(layer)))
+	for ch := range t.subscribers { // notify all subscribers
+
+		// non-blocking notify
+		select {
+		case ch <- layer:
+			continue
+		default:
+			missedTicks++ // count subscriber that missed tick
+			continue
+		}
+	}
+
+	// update last ticked layer
+	t.lastTickedLayer = layer
+
+	if missedTicks > 0 {
+		t.log.With().Error("missed ticks for layer",
+			log.LayerId(uint64(layer)), log.Int("missed_count", missedTicks))
+		t.m.Unlock()
+		return missedTicks, errMissedTicks
+	}
+
+	t.m.Unlock()
+	return 0, nil
+}
+
+// TimeSinceLastTick returns the duration passed since the last layer that we ticked
+// note: the call is not lock-protected
+func (t *Ticker) timeSinceLastTick() time.Duration {
+	timeOfLastTick := t.conv.LayerToTime(t.conv.TimeToLayer(t.clock.Now()))
```

Comment:
woun't this always be 0 since you are using time.now in both ends of the equation? shouldn't you use `lastTickedLayer` here somewhere?

Reply:
It will be zero iff the Now is exactly on the layer time.